core/src/actors/json_rpc/server.rs

```diff
@@ -0,0 +1,139 @@
+use tokio::{codec::FramedRead, io::AsyncRead};
+use witnet_config::config::Config;
+
+/// JSON RPC server
+#[derive(Default)]
+pub struct JsonRpcServer {
+    /// Server address
+    server_addr: Option<SocketAddr>,
+    /// Open connections, stored as instances of the `JsonRpc` actor
+    open_connections: HashSet<Addr<JsonRpc>>,
+    /// JSON-RPC methods
+    // Stored as an `Rc` to avoid creating a new handler for each connection
+    jsonrpc_io: Option<Rc<IoHandler<()>>>,
+}
+
+impl JsonRpcServer {
+    /// Method to process the configuration received from the ConfigManager
+    fn process_config(&mut self, ctx: &mut <Self as Actor>::Context, config: &Config) {
+        let enabled = config.jsonrpc.enabled;
+
+        // Do not start the server if enabled = false
+        if !enabled {
+            debug!("JSON RPC server actor not started because of config!");
+            ctx.stop();
+            return;
+        }
+
+        debug!("JSON RPC server actor has been started!");
+        let server_addr = config.jsonrpc.server_address;
+        self.server_addr = Some(server_addr);
+        // Create and store the JSON-RPC method handler
+        let jsonrpc_io = jsonrpc_io_handler();
+        self.jsonrpc_io = Some(Rc::new(jsonrpc_io));
+
+        // Bind TCP listener to this address
+        // FIXME(#176): running `yes | nc 127.0.0.1 1234` freezes the entire actor system
+        let listener = match TcpListener::bind(&server_addr) {
+            Ok(listener) => listener,
+            Err(e) => {
+                // Shutdown the entire system on error
+                // For example, when the server_addr is already in use
+                // FIXME(#72): gracefully stop the system?
+                error!("Could not start JSON-RPC server: {:?}", e);
+                panic!("Could not start JSON-RPC server: {:?}", e);
+            }
+        };
+
+        // Add message stream which will return a InboundTcpConnect for each incoming TCP connection
+        ctx.add_message_stream(
+            listener
+                .incoming()
+                .map_err(|_| ())
+                .map(InboundTcpConnect::new),
+        );
+
+        debug!("Started JSON RPC server at {}", server_addr);
+    }
+
+    fn add_connection(&mut self, parent: Addr<JsonRpcServer>, stream: TcpStream) {
+        debug!(
+            "Add session (currently {} open connections)",
+            1 + self.open_connections.len()
+        );
+
+        // Get a reference to the JSON-RPC method handler
+        let jsonrpc_io = Rc::clone(self.jsonrpc_io.as_ref().unwrap());
+
+        // Create a new `JsonRpc` actor which will listen to this stream
+        let addr = JsonRpc::create(|ctx| {
+            let (r, w) = stream.split();
+            JsonRpc::add_stream(FramedRead::new(r, NewLineCodec), ctx);
+            JsonRpc {
+                framed: FramedWrite::new(w, NewLineCodec, ctx),
+                parent,
+                jsonrpc_io,
+            }
```

Comment:
I'm not sure if `NewLineCodec` is suitable here (or I'm understanding it use correctly), what happens if the JSON RPC message is sent without minification (that is, containing newlines)?

Reply:
Yes, we assume that the JSON RPC message will be sent in one line, otherwise it doesn't work. We should add a warning to the documentation.